### ISSUE TITLE
Method implementation

### ISSUE THREAD

**Issue by jakebolewski** [JuliaParallel/DistributedArrays.jl#4]
### object identity and equality
- [ ] copy
- [ ] deepcopy
- [ ] hash
- [ ] identity
- [ ] isbits
- [ ] isequal
- [ ] isimmutable
- [ ] isless
- [ ] ifelse
- [ ] lexless
- [ ] lexcmp
- [ ] object_id
- [ ] sizeof

### collections
- [ ] all!
- [ ] all
- [ ] any!
- [ ] any
- [ ] collect
- [ ] complement!
- [ ] complement
- [ ] contains
- [ ] count
- [ ] delete!
- [ ] deleteat!
- [ ] eltype
- [ ] empty!
- [ ] endof
- [ ] filter!
- [ ] filter
- [ ] foldl
- [ ] foldr
- [ ] get
- [ ] get!
- [ ] getindex
- [ ] getkey
- [ ] haskey
- [ ] in
- [ ] intersect!
- [ ] intersect
- [ ] isempty
- [ ] issubset
- [ ] keys
- [ ] length
- [ ] map!
- [ ] map
- [ ] mapfoldl
- [ ] mapfoldr
- [ ] mapreduce
- [ ] mapreducedim
- [ ] merge!
- [ ] merge
- [ ] reduce
- [ ] setdiff!
- [ ] setdiff
- [ ] setindex!
- [ ] similar
- [ ] sizehint!
- [ ] splice!
- [ ] symdiff!
- [ ] symdiff
- [ ] union!
- [ ] union
- [ ] unique
- [ ] values

### iteration
- [ ] done
- [ ] enumerate
- [ ] next
- [ ] start
- [ ] zip
- [ ] rest
- [ ] countfrom
- [ ] take
- [ ] drop
- [ ] cycle
- [ ] repeated

### arrays
- [ ] bitbroadcast
- [ ] broadcast!
- [ ] broadcast!_function
- [ ] broadcast
- [ ] broadcast_function
- [ ] broadcast_getindex
- [ ] broadcast_setindex!
- [ ] cartesianmap
- [ ] cat
- [ ] cell
- [ ] checkbounds
- [ ] circshift
- [ ] colon
- [ ] conj!
- [ ] copy!
- [ ] cummax
- [ ] cummin
- [ ] cumprod
- [ ] cumprod!
- [ ] cumsum
- [ ] cumsum!
- [ ] cumsum_kbn
- [ ] eachindex
- [ ] extrema
- [ ] fill!
- [ ] fill
- [ ] find
- [ ] findfirst
- [ ] findlast
- [ ] findin
- [ ] findmax
- [ ] findmin
- [ ] findn
- [ ] findnext
- [ ] findprev
- [ ] findnz
- [ ] first
- [ ] flipdim
- [ ] gradient
- [ ] hcat
- [ ] hvcat
- [ ] ind2sub
- [ ] indexin
- [ ] indmax
- [ ] indmin
- [ ] invperm
- [ ] ipermute!
- [ ] ipermutedims
- [ ] isperm
- [ ] issorted
- [ ] last
- [ ] linrange
- [ ] linspace
- [ ] logspace
- [ ] mapslices
- [ ] max
- [ ] maxabs
- [ ] maxabs!
- [ ] maximum!
- [ ] maximum
- [ ] min
- [ ] minabs
- [ ] minabs!
- [ ] minimum!
- [ ] minimum
- [ ] minmax
- [ ] ndims
- [ ] nnz
- [ ] nonzeros
- [ ] nthperm!
- [ ] nthperm
- [ ] countnz
- [ ] ones
- [ ] parent
- [ ] parentindexes
- [ ] partitions
- [ ] permutations
- [ ] permute!
- [ ] permutedims
- [ ] permutedims!
- [ ] prod!
- [ ] prod
- [ ] promote_shape
- [ ] randcycle
- [ ] randperm
- [ ] randsubseq!
- [ ] randsubseq
- [ ] range
- [ ] reducedim
- [ ] repmat
- [ ] reshape
- [ ] reverse!
- [ ] reverse
- [ ] rot180
- [ ] rotl90
- [ ] rotr90
- [ ] searchsorted
- [ ] searchsortedfirst
- [ ] searchsortedlast
- [ ] select!
- [ ] select
- [ ] shuffle
- [ ] shuffle!
- [ ] size
- [ ] slice
- [ ] slicedim
- [ ] sort!
- [ ] sort
- [ ] sortcols
- [ ] sortperm
- [ ] sortperm!
- [ ] sortrows
- [ ] squeeze
- [ ] step
- [ ] stride
- [ ] strides
- [ ] sub
- [ ] sub2ind
- [ ] sum!
- [ ] sum
- [ ] sumabs!
- [ ] sumabs
- [ ] sumabs2!
- [ ] sumabs2
- [ ] sum_kbn
- [ ] vcat
- [ ] vec
- [ ] zeros

### linear algebra
- [ ] bkfact!
- [ ] bkfact
- [ ] blas_set_num_threads
- [ ] blkdiag
- [ ] chol
- [ ] cholfact!
- [ ] cholfact
- [ ] cond
- [ ] condskeel
- [ ] cross
- [ ] ctranspose!
- [ ] ctranspose
- [ ] det
- [ ] diag
- [ ] diagind
- [ ] diagm
- [ ] diff
- [ ] dot
- [ ] eig
- [ ] eigfact!
- [ ] eigfact
- [ ] eigmax
- [ ] eigmin
- [ ] eigs
- [ ] eigvals
- [ ] eigvals!
- [ ] eigvecs
- [ ] expm
- [ ] eye
- [ ] factorize
- [ ] givens
- [ ] hessfact!
- [ ] hessfact
- [ ] ishermitian
- [ ] isposdef!
- [ ] isposdef
- [ ] issym
- [ ] istril
- [ ] istriu
- [ ] kron
- [ ] ldltfact
- [ ] ldltfact!
- [ ] linreg
- [ ] logdet
- [ ] lu
- [ ] lufact!
- [ ] lufact
- [ ] lyap
- [ ] norm
- [ ] nullspace
- [ ] ordschur!
- [ ] ordschur
- [ ] peakflops
- [ ] pinv
- [ ] qr
- [ ] qrfact!
- [ ] qrfact
- [ ] rank
- [ ] scale!
- [ ] scale
- [ ] schur
- [ ] schurfact!
- [ ] schurfact
- [ ] sqrtm
- [ ] svd
- [ ] svdfact!
- [ ] svdfact
- [ ] svds
- [ ] svdvals!
- [ ] svdvals
- [ ] sylvester
- [ ] trace
- [ ] transpose!
- [ ] transpose
- [ ] tril!
- [ ] tril
- [ ] triu!
- [ ] triu
- [ ] vecnorm

### sparse
- [ ] etree
- [ ] full
- [ ] issparse
- [ ] sparse
- [ ] sparsevec
- [ ] spdiagm
- [ ] speye
- [ ] spones
- [ ] sprand
- [ ] sprandbool
- [ ] sprandn
- [ ] spzeros
- [ ] symperm
- [ ] rowvals
- [ ] nzrange

# Operators
- [ ] !
- [ ] !=
- [ ] ≠
- [ ] !==
- [ ] ≡
- [ ] ≢
- [ ] $
- [ ] %
- [ ] &
- [ ] *
- [ ] +
- [ ] -
- [ ] .!=
- [ ] .≠
- [ ] .+
- [ ] .-
- [ ] .*
- [ ] ./
- [ ] .<
- [ ] .<=
- [ ] .≤
- [ ] .==
- [ ] .>
- [ ] .>=
- [ ] .≥
- [ ] .\
- [ ] .^
- [ ] /
- [ ] //
- [ ] .//
- [ ] <
- [ ] <<
- [ ] <=
- [ ] ≤
- [ ] ==
- [ ] >
- [ ] >=
- [ ] ≥
- [ ] >>
- [ ] >>>
- [ ] \
- [ ] ^
- [ ] |
- [ ] ~
- [ ] ÷
- [ ] =>
- [ ] A_ldiv_B!
- [ ] A_ldiv_Bc
- [ ] A_ldiv_Bt
- [ ] A_mul_B!
- [ ] A_mul_Bc
- [ ] A_mul_Bc!
- [ ] A_mul_Bt
- [ ] A_mul_Bt!
- [ ] A_rdiv_Bc
- [ ] A_rdiv_Bt
- [ ] Ac_ldiv_B
- [ ] Ac_ldiv_Bc
- [ ] Ac_mul_B
- [ ] Ac_mul_B!
- [ ] Ac_mul_Bc
- [ ] Ac_mul_Bc!
- [ ] Ac_rdiv_B
- [ ] Ac_rdiv_Bc
- [ ] At_ldiv_B
- [ ] At_ldiv_Bt
- [ ] At_mul_B
- [ ] At_mul_B!
- [ ] At_mul_Bt
- [ ] At_mul_Bt!
- [ ] At_rdiv_B
- [ ] At_rdiv_Bt

### scalar math
- [ ] abs
- [ ] abs2
- [ ] acos
- [ ] acosd
- [ ] acosh
- [ ] acot
- [ ] acotd
- [ ] acoth
- [ ] acsc
- [ ] acscd
- [ ] acsch
- [ ] angle
- [ ] asec
- [ ] asecd
- [ ] asech
- [ ] asin
- [ ] asind
- [ ] asinh
- [ ] atan
- [ ] atan2
- [ ] atand
- [ ] atanh
- [ ] big
- [ ] binomial
- [ ] bswap
- [ ] cbrt
- [ ] ceil
- [ ] cis
- [ ] clamp
- [ ] cld
- [ ] cmp
- [ ] combinations
- [ ] complex
- [ ] conj
- [ ] copysign
- [ ] cos
- [ ] cosc
- [ ] cosd
- [ ] cosh
- [ ] cospi
- [ ] cot
- [ ] cotd
- [ ] coth
- [ ] count_ones
- [ ] count_zeros
- [ ] csc
- [ ] cscd
- [ ] csch
- [ ] dawson
- [ ] deg2rad
- [ ] den
- [ ] digamma
- [ ] div
- [ ] divrem
- [ ] eps
- [ ] erf
- [ ] erfc
- [ ] erfcinv
- [ ] erfcx
- [ ] erfi
- [ ] erfinv
- [ ] exp
- [ ] exp10
- [ ] exp2
- [ ] expm1
- [ ] exponent
- [ ] factor
- [ ] factorial
- [ ] fld
- [ ] fldmod
- [ ] flipsign
- [ ] float
- [ ] floor
- [ ] fma
- [ ] frexp
- [ ] gamma
- [ ] gcd
- [ ] gcdx
- [ ] hex2num
- [ ] hypot
- [ ] imag
- [ ] inv
- [ ] invdigamma
- [ ] invmod
- [ ] isapprox
- [ ] iseltype
- [ ] iseven
- [ ] isfinite
- [ ] isinf
- [ ] isinteger
- [ ] isnan
- [ ] isodd
- [ ] ispow2
- [ ] isprime
- [ ] isqrt
- [ ] isreal
- [ ] issubnormal
- [ ] lcm
- [ ] ldexp
- [ ] leading_ones
- [ ] leading_zeros
- [ ] lfact
- [ ] lgamma
- [ ] log
- [ ] log10
- [ ] log1p
- [ ] log2
- [ ] maxintfloat
- [ ] mod
- [ ] mod1
- [ ] modf
- [ ] mod2pi
- [ ] muladd
- [ ] nextfloat
- [ ] nextpow
- [ ] nextpow2
- [ ] nextprod
- [ ] num
- [ ] num2hex
- [ ] one
- [ ] powermod
- [ ] prevfloat
- [ ] prevpow
- [ ] prevpow2
- [ ] prevprod
- [ ] primes
- [ ] rad2deg
- [ ] rationalize
- [ ] real
- [ ] realmax
- [ ] realmin
- [ ] reim
- [ ] reinterpret
- [ ] rem
- [ ] rem1
- [ ] round
- [ ] sec
- [ ] secd
- [ ] sech
- [ ] sign
- [ ] signbit
- [ ] signed
- [ ] signif
- [ ] significand
- [ ] sin
- [ ] sinc
- [ ] sind
- [ ] sinh
- [ ] sinpi
- [ ] sqrt
- [ ] tan
- [ ] tand
- [ ] tanh
- [ ] trailing_ones
- [ ] trailing_zeros
- [ ] trigamma
- [ ] trunc
- [ ] unsafe_trunc
- [ ] typemax
- [ ] typemin
- [ ] unsigned
- [ ] widemul
- [ ] zero

# specfun
- [ ] airy
- [ ] airyai
- [ ] airyaiprime
- [ ] airybi
- [ ] airybiprime
- [ ] airyprime
- [ ] airyx
- [ ] besselh
- [ ] besseli
- [ ] besselix
- [ ] besselj
- [ ] besselj0
- [ ] besselj1
- [ ] besseljx
- [ ] besselk
- [ ] besselkx
- [ ] bessely
- [ ] bessely0
- [ ] bessely1
- [ ] besselyx
- [ ] beta
- [ ] eta
- [ ] hankelh1
- [ ] hankelh1x
- [ ] hankelh2
- [ ] hankelh2x
- [ ] lbeta
- [ ] polygamma
- [ ] zeta

### random numbers
- [ ] rand!
- [ ] rand
- [ ] randn!
- [ ] randn
- [ ] randexp!
- [ ] randexp
- [ ] srand
- [ ] bitrand

### statistics
- [ ] cor
- [ ] cov
- [ ] hist
- [ ] hist!
- [ ] hist2d
- [ ] hist2d!
- [ ] histrange
- [ ] mean!
- [ ] mean
- [ ] median!
- [ ] median
- [ ] middle
- [ ] midpoints
- [ ] quantile!
- [ ] quantile
- [ ] std
- [ ] stdm
- [ ] var
- [ ] varm

### signal processing
- [ ] bfft!
- [ ] bfft
- [ ] brfft
- [ ] conv
- [ ] conv2
- [ ] dct!
- [ ] dct
- [ ] deconv
- [ ] fft!
- [ ] fft
- [ ] fftshift
- [ ] filt
- [ ] filt!
- [ ] idct!
- [ ] idct
- [ ] ifft!
- [ ] ifft
- [ ] ifftshift
- [ ] irfft
- [ ] plan_bfft!
- [ ] plan_bfft
- [ ] plan_brfft
- [ ] plan_dct!
- [ ] plan_dct
- [ ] plan_fft!
- [ ] plan_fft
- [ ] plan_idct!
- [ ] plan_idct
- [ ] plan_ifft!
- [ ] plan_ifft
- [ ] plan_irfft
- [ ] plan_rfft
- [ ] rfft
- [ ] xcorr

**Comment by jiahao** [comment]
At a first pass we can skip most of the vectorized functions - i.e. all the specfuns and most of the scalar mathematical functions.

**Comment by ViralBShah** [comment]
Or implement them with macros and conquer much of this list. :-)

I love this list. I am sure @alanedelman loves it too.

**Comment by ViralBShah** [comment]
621 functions. 16 done. Only 605 to go. :-)

Cc: @amitmurthy @tanmaykm @alanedelman in case you guys haven't seen this.

**Comment by jakebolewski** [comment]
Some of these will be hard / impossible to do in parallel.  My favorite is `==` if we want to support the same semantics as `AbstractArrays` now in base.

**Comment by ViralBShah** [comment]
Yes, of course! 

What is the exact issue with `==` ? Perhaps should be a separate issue.

**Comment by jiahao** [comment]
@ViralBShah I agree == (with hash) should be their own issue. The challenge is how to hash DArrays in parallel so that they can test equal under == regardless of distribution. The only way we can guarantee that now for A==B is to redistribute all the data onto a single process and do the hash locally for A and B.

**Comment by jakebolewski** [comment]
@tanmaykm, @amitmurthy it would be great if you could check off functions as they are added.  I  believe anyone that is an owner of the repo can edit the check list.

**Comment by alanedelman** [comment]
don't we have svd and a few other scalapack functions now?

**Comment by ViralBShah** [comment]
We should probably separate out linear algebra related function checklist into scalapack.jl or elemental.jl, and have this issue track only DArrays.

**Comment by alanedelman** [comment]
but i'd like this to work with darrays :-)

**Comment by andreasnoack** [comment]
It will, but the dependency order right now is that ScaLAPACK.jl depends on DistributedArrays.jl and, therefore, the DArray+ScaLAPACK functionality is in ScaLAPACK.jl.

**Comment by andreasnoack** [comment]
@alanedelman Just to avoid misunderstanding here. You can do
```julia
julia> using MPI

julia> man = MPIManager(np = 4);

julia> addprocs(man);

julia> using ScaLAPACK
Warning: Method definition defaultdistTuple{Int64,Int64} in module DistributedArrays at /Users/andreasnoack/.julia/v0.4/DistributedArrays/src/DistributedArrays.jl:147 overwritten in module ScaLAPACK at /Users/andreasnoack/.julia/v0.4/ScaLAPACK/src/ScaLAPACK.jl:13.
	From worker 4:	Warning: Method definition defaultdistTuple{Int64,Int64} in module DistributedArrays at /Users/andreasnoack/.julia/v0.4/DistributedArrays/src/DistributedArrays.jl:147 overwritten in module ScaLAPACK at /Users/andreasnoack/.julia/v0.4/ScaLAPACK/src/ScaLAPACK.jl:13.
	From worker 3:	Warning: Method definition defaultdistTuple{Int64,Int64} in module DistributedArrays at /Users/andreasnoack/.julia/v0.4/DistributedArrays/src/DistributedArrays.jl:147 overwritten in module ScaLAPACK at /Users/andreasnoack/.julia/v0.4/ScaLAPACK/src/ScaLAPACK.jl:13.
	From worker 2:	Warning: Method definition defaultdistTuple{Int64,Int64} in module DistributedArrays at /Users/andreasnoack/.julia/v0.4/DistributedArrays/src/DistributedArrays.jl:147 overwritten in module ScaLAPACK at /Users/andreasnoack/.julia/v0.4/ScaLAPACK/src/ScaLAPACK.jl:13.
	From worker 5:	Warning: Method definition defaultdistTuple{Int64,Int64} in module DistributedArrays at /Users/andreasnoack/.julia/v0.4/DistributedArrays/src/DistributedArrays.jl:147 overwritten in module ScaLAPACK at /Users/andreasnoack/.julia/v0.4/ScaLAPACK/src/ScaLAPACK.jl:13.

julia> @everywhere using DistributedArrays

julia> A = drandn((1000, 1000), man);

julia> B = convert(Array, A);

julia> @time valsA = svdvals(A);
elapsed time: 1.382340321 seconds (37 MB allocated, 1.53% gc time in 1 pauses with 0 full sweep)

julia> @time valsA = svdvals(A);
elapsed time: 0.323082764 seconds (1 MB allocated)

julia> @time valsB = svdvals(B);
elapsed time: 1.0682854 seconds (11 MB allocated, 0.48% gc time in 1 pauses with 0 full sweep)

julia> @time valsB = svdvals(B);
elapsed time: 0.998921006 seconds (8 MB allocated)```
The warnings will go away eventually.

**Comment by alanedelman** [comment]
I was kind of sad that now that we are about to have the "dot"  (Pronounced broadcast
but I don't like that name, I prefer pointwise or just point)  why we don't have these
functions after all these years.

Why can't we have all the operators and all the dot operators easily?

**Comment by andreasnoack** [comment]
We should have all the dot stuff now but it only works on Julia 0.6 and master. Somebody just needs to go through the list and check. Some of the items are also deprecated now.

**Comment by andreasnoack** [comment]
@alanedelman I just checked and all the dot stuff works now for DArrays. However, they are not independent functions anymore but just syntax so I've removed them from the list. The important detail is that `broadcast` works.

**Comment by alanedelman** [comment]
where are we on cumsum, cumprod, and accumulate ?

**Comment by andreasnoack** [comment]
Still some work to be done there.

**Comment by alanedelman** [comment]
would love to bring this issue back to life

**Comment by montyvesselinov** [comment]
me too

**Comment by ViralBShah** [comment]
@christopher-dG Just bringing this list to your attention. Maybe we should discuss this today?

**Comment by alanedelman** [comment]
+1
